File: web/src/services/notification.service.ts
import Fetch from "@/lib/fetch";
import type {
  ApiResponse,
  Notification,
  PaginationParams,
} from "@/types";

class NotificationService {
  /**
   * Get all notifications for the authenticated user
   */
  async getNotifications(params?: PaginationParams): Promise<ApiResponse<Notification[]>> {
    const queryParams = new URLSearchParams();
    if (params?.limit) queryParams.append("limit", params.limit.toString());
    if (params?.offset) queryParams.append("offset", params.offset.toString());

    const response = await Fetch.get<ApiResponse<Notification[]>>(
      `/notifications?${queryParams.toString()}`
    );
    return response.data;
  }

  /**
   * Get unread notification count
   */
  async getUnreadCount(): Promise<ApiResponse<{ count: number }>> {
    const response = await Fetch.get<ApiResponse<{ count: number }>>(
      "/notifications/unread-count"
    );
    return response.data;
  }

  /**
   * Mark a notification as read
   */
  async markAsRead(id: number): Promise<ApiResponse<null>> {
    const response = await Fetch.put<ApiResponse<null>>(
      `/notifications/${id}/read`,
      {}
    );
    return response.data;
  }

  /**
   * Mark all notifications as read
   */
  async markAllAsRead(): Promise<ApiResponse<null>> {
    const response = await Fetch.put<ApiResponse<null>>(
      "/notifications/read-all",
      {}
    );
    return response.data;
  }

  /**
   * Delete a notification
   */
  async deleteNotification(id: number): Promise<ApiResponse<null>> {
    const response = await Fetch.delete<ApiResponse<null>>(`/notifications/${id}`);
    return response.data;
  }
}

export default new NotificationService();

File: modules/notification/domain/service/notification_service.go
package service

import (
	"Ravit/modules/notification/domain/entity"
	"Ravit/modules/notification/domain/repository"
	"context"
	"errors"
)

// Errors
var (
	ErrNotificationNotFound = errors.New("notification not found")
)

// NotificationService handles notification domain logic
type NotificationService struct {
	notificationRepo repository.NotificationRepository
}

// NewNotificationService creates a new notification service
func NewNotificationService(notificationRepo repository.NotificationRepository) *NotificationService {
	return &NotificationService{
		notificationRepo: notificationRepo,
	}
}

// CreateNotification creates a new notification
func (s *NotificationService) CreateNotification(ctx context.Context, notification *entity.Notification) error {
	return s.notificationRepo.Create(ctx, notification)
}

// GetNotificationByID gets a notification by ID
func (s *NotificationService) GetNotificationByID(ctx context.Context, id uint) (*entity.Notification, error) {
	notification, err := s.notificationRepo.FindByID(ctx, id)
	if err != nil {
		return nil, ErrNotificationNotFound
	}
	return notification, nil
}

// GetNotificationsByUserID gets all notifications for a user
func (s *NotificationService) GetNotificationsByUserID(ctx context.Context, userID uint, limit, offset int) ([]*entity.Notification, error) {
	return s.notificationRepo.FindByUserID(ctx, userID, limit, offset)
}

// GetUnreadCount gets the count of unread notifications for a user
func (s *NotificationService) GetUnreadCount(ctx context.Context, userID uint) (int64, error) {
	return s.notificationRepo.CountUnreadByUserID(ctx, userID)
}

// MarkAsRead marks a notification as read
func (s *NotificationService) MarkAsRead(ctx context.Context, id uint) error {
	return s.notificationRepo.MarkAsRead(ctx, id)
}

// MarkAllAsRead marks all notifications for a user as read
func (s *NotificationService) MarkAllAsRead(ctx context.Context, userID uint) error {
	return s.notificationRepo.MarkAllAsRead(ctx, userID)
}

// DeleteNotification deletes a notification
func (s *NotificationService) DeleteNotification(ctx context.Context, id uint) error {
	return s.notificationRepo.Delete(ctx, id)
}
